gerrytools/geometry/dissolve.py

```diff
@@ -33,3 +37,144 @@ def dissolve(
         geometries = geometries.reset_index()
 
     return geometries
+
+
+def hierarchical_block_dissolve(
+    state,
+    block_assignment: DataFrame,
+    district_col: str,
+    area_consistency_tol: float = 1e-4
+) -> Tuple[GeoDataFrame, Counter]:
+    """Hierarchically dissolves Census blocks into polygons by assignment.
+    Dissolving blocks into districts for an entire state is a notoriously
+    expensive operation. However, real-life plans tend to preserve block group
+    and county boundaries in many places. This suggests a dissolving algorithm
+    that leverages the Census hierarchy (blocks -> block groups -> tracts ->
+    counties) and merges the largest whole polygons possible---for instance,
+    if a whole county is contained within a district, we use the precomputed
+    county polygon rather than dissolving its constituent blocks.
+    `GeoDataFrame`s of Census data must have the same CRS and Census vintage.
+    They must be indexed by their vintage-specific `GEOID` (e.g. `GEOID10`
+    or `GEOID20`).
+    Args:
+        state (State): us.states.STATE
+        block_assignment: A DataFrame with block ids and a district column.
+        district_col: Name of column with plan districts.
+        area_consistency_tol: Relative tolerance for area consistency check.
+    Raises:
+        ValueError: If the CRSes of the `GeoDataFrame`s do not match
+            or `block_gdf` is unspecified.
+        DisolveError: If areas are not approximately consistent
+            between block unions and hierarchical unions.
+    Returns:
+        A tuple containing a `GeoDataFrame` of dissolved district polygons
+        (indexed by label) and counts of polygons used from each level
+        in the hierarchy.
+    """
+    state_abbr = str(state.abbr).lower()
+    base = "http://data.mggg.org.s3-website.us-east-2.amazonaws.com/census-2020"
+
+    block_gdf = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_block.zip").set_index("GEOID20")
+    bg_gdf = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_bg.zip").to_crs(block_gdf.crs).set_index("GEOID20")
+    tract_gdf = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_tract.zip").to_crs(block_gdf.crs).set_index("GEOID20")
+    county_gdf = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_county.zip").to_crs(block_gdf.crs).set_index("GEOID20")
```

Comment:
This is inefficient if this function is called many times, as the datasets will be redownloaded at every call (to my knowledge, there's no global cache). Instead, I think we should write a wrapper function that grabs all these datasets and passes them in as one object. I could imagine defining something like this:

```python
import us
import geopandas
from dataclasses import dataclass

S3_CENSUS_2020_BASE = "http://data.mggg.org.s3-website.us-east-2.amazonaws.com/census-2020"


@dataclass
class StateHierarchy:
    state: us.states.STATE
    blocks: Optional[gpd.GeoDataFrame]
    block_groups: Optional[gpd.GeoDataFrame]
    tracts: Optional[gpd.GeoDataFrame]
    counties: Optional[gpd.GeoDataFrame]

    @staticmethod
    def from_s3(cls, state: us.states.STATE, base: str = S3_CENSUS_2020_BASE) -> 'StateHierarchy':
        state_abbr = str(state.abbr).lower()
        blocks = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_block.zip").set_index("GEOID20")
        block_groups = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_bg.zip").to_crs(block_gdf.crs).set_index("GEOID20")
        tracts = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_tract.zip").to_crs(block_gdf.crs).set_index("GEOID20")
        counties = gpd.read_file(f"{base}/{state_abbr}/{state_abbr}_county.zip").to_crs(block_gdf.crs).set_index("GEOID20")

        return cls(
            state=state,
            blocks=blocks,
            block_groups=block_groups,
            tracts=tracts,
            counties=counties
        )